onnxruntime/core/optimizer/compute_optimizer/upstream_reshape_actors.cc

```diff
@@ -0,0 +1,287 @@
+  //
+  // The dims are compatible as long as it meets one of the following conditions:
+  // 1. The first two dims do not exist (e.g. input shape is [1024] and output shape is [2, 16, 1024])
+  // 2. The first two dims exist, and are equal (e.g. input shape is [2, 16, 1024] and output shape is [2, 16, 1024])
+  // 3. [Not Supported] The first two dims exist, and the first dim is 1 (e.g. input shape is [1, 16, 1024] and output
+  //    shape is [2, 16, 1024])
+  // 4. [Not Supported] The first two dims exist, and the second dim is 1 (e.g. input shape is [2, 1, 1024] and output
+  //    shape is [2, 16, 1024])
+  // 5. [Not Supported] The first dim does not exist, and the second dim exist (can be 1 or not).
+  for (int input_idx = 0; input_idx < static_cast<int>(current_node.InputDefs().size()); ++input_idx) {
+    auto input_shape = current_node.InputDefs()[input_idx]->Shape();
+
+    auto [success, ret] = CompareInputShapeWithOutputShape(data_input_arg->Shape(),
+                                                           input_shape);
+
+    if (!success) {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": reshape's data input rank < passthrough node's input rank");
+      return false;
+    }
+
+    if (ret.size() < 2) {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": full broadcasted shape has dim < 2.");
+      return false;
+    }
+
+    all_input_cmp_rets[input_idx] = ret;
+
+    if (ret[0] == DimCompareRet::NotExist && ret[1] == DimCompareRet::NotExist) {
+      // Don't need to propagate Reshape since the two leading dims to flatten do not ext.
+      LOG_DEBUG_INFO(logger, "In SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": skip propagating for the input because the merged dims does not exist.");
+      continue;
+    } else if (ret[0] == DimCompareRet::Equal && ret[1] == DimCompareRet::Equal) {
+      propagate_input_indices.push_back(input_idx);
+    } else {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": not supported cases for two leading dims check.");
+      return false;
+    }
+
+    // All other dims should be equal
+    for (int dim_index = 2; dim_index < static_cast<int>(ret.size()); ++dim_index) {
+      if (ret[dim_index] != DimCompareRet::Equal) {
+        LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                   ": unflatten dims should all be equal to full broadcast shape.");
+        return false;
+      }
+    }
+  }
+
+  if (AreAllOutputShapesEqual) {
+    // Make sure once Reshape is moved before the target node, all its outputs can be correctly reshaped.
```

Comment:
why need all output shapes equal?

Reply:
Once a Reshape node passed through a node (we called current_node) before it, we handle the input propagation, e.g. insert Reshape node for those inputs that requires propagation. While we need also make sure after that, if current_node has other outputs that are not used by the initial Reshape node, their shape should also be adapted. Currently, we don't have any case for multiple outputs, so we don't implement the output adaption logic for other outputs, instead we restrict it during PreCheck. 

If we see a demand later to handle this case, we can optionally create a new Upstream actor inheriting from SimplePointwiseReshapeActor<false> (which makes template type AreAllOutputShapesEqual to be false), and then do the output adaption in the PostProcess .